src/cluster-client/src/client.rs

```diff
@@ -0,0 +1,216 @@
+        Self {
+            envd: envd.try_into().unwrap(),
+            replica,
+        }
+    }
+
+    pub fn envd(&self) -> NonZeroI64 {
+        self.envd
+    }
+
+    pub fn replica(&self) -> u64 {
+        self.replica
+    }
+}
+
+impl std::fmt::Display for ClusterStartupEpoch {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let Self { envd, replica } = self;
+        write!(f, "({envd}, {replica})")
+    }
+}
+
+impl PartialOrd for ClusterStartupEpoch {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for ClusterStartupEpoch {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        let Self { envd, replica } = self;
+        let Self {
+            envd: other_envd,
+            replica: other_replica,
+        } = other;
+        (envd, replica).cmp(&(other_envd, other_replica))
+    }
+}
+
+/// Configuration of the cluster we will spin up
+#[derive(Arbitrary, Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
+pub struct TimelyConfig {
+    /// Number of per-process worker threads
+    pub workers: usize,
+    /// Identity of this process
+    pub process: usize,
+    /// Addresses of all processes
+    pub addresses: Vec<String>,
+    /// The amount of effort to be spent on arrangement compaction during idle times.
+    ///
+    /// See `differential_dataflow::Config::idle_merge_effort`.
+    pub idle_arrangement_merge_effort: u32,
+}
+
+impl RustType<ProtoTimelyConfig> for TimelyConfig {
+    fn into_proto(&self) -> ProtoTimelyConfig {
+        ProtoTimelyConfig {
+            workers: self.workers.into_proto(),
+            addresses: self.addresses.into_proto(),
+            process: self.process.into_proto(),
+            idle_arrangement_merge_effort: self.idle_arrangement_merge_effort,
+        }
+    }
+
+    fn from_proto(proto: ProtoTimelyConfig) -> Result<Self, TryFromProtoError> {
+        Ok(Self {
+            process: proto.process.into_rust()?,
+            workers: proto.workers.into_rust()?,
+            addresses: proto.addresses.into_rust()?,
+            idle_arrangement_merge_effort: proto.idle_arrangement_merge_effort,
+        })
+    }
+}
+
+impl TimelyConfig {
+    pub fn split_command(&self, parts: usize) -> Vec<Self> {
+        (0..parts)
+            .into_iter()
+            .map(|part| TimelyConfig {
+                process: part,
+                ..self.clone()
+            })
+            .collect()
+    }
+}
+
+/// Specifies the location of a cluster replica.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct ClusterReplicaLocation {
+    /// The network addresses of the cluster control endpoints for each process in
+    /// the replica.
+    pub ctl_addrs: Vec<String>,
+    /// The network addresses of the dataflow (Timely) endpoints for
+    /// each process in the replica.
+    pub dataflow_addrs: Vec<String>,
```

Comment:
I know this is not your PR, but I had to look up what exactly these mean. Might be good to document that as a follow-up. My understanding is:
 - ctl_addrs is for the controller talking to the "cluster"
 - dataflow_addrs is for internal networking, such that the different processes that make up a cluster can find and talk to each other
Is that correct?

Reply:
I will look at this as a followup!